determined-ai/determined
# feat: add new endpoints to change log retention for exps and trials [DET-9995]

## Description



Added new API Endpoints, CLI commands and WebUI elements to change log retention for experiments and trials

## Test Plan


- Added integration tests to test the API

For CLI - 
Experiment:
1. Create an experiment.
2. Wait for the experiment to complete.
3. Run `det e set log-retention-days <exp-id> --days 50`
4. Check the tasks table in the database to see if the associated tasks for the experiment have updated the log retention days to 50.
5.  Run `det e set log-retention-days <exp-id> --days none --forever`
6. Check the tasks table in the database to see if the associated tasks for the experiment have updated the log retention days to -1.

Trial:
1. Create an experiment.
2. Wait for the experiment to complete.
3. Run `det t set log-retention-days <trial-id> --days 50` for any trial in the experiment.
4. Check the tasks table in the database to see if the associated tasks for the trial have updated the log retention days to 50.
5.  Run `det t set log-retention-days <trial-id> --days none --forever`
6. Check the tasks table in the database to see if the associated tasks for the trial have updated the log retention days to -1.

For API - 
1. Create an experiment.
2. Wait for the experiment to complete.
3. Use the following API endpoint:

> curl -X 'PUT' \
>   'http://localhost:8080/api/v1/experiments/1/retain_logs' \
>   -H 'accept: application/json' \
>   -H 'Content-Type: application/json' \
>   -d '{
>   "experimentId": 1,
>   "numDays": 40
> }'

Check the tasks table in the database to see if the associated tasks for the experiment have updated the log retention days to 40.

> curl -X 'PUT' \
>  'http://localhost:8080/api/v1/trials/1/retain_logs' \
>   -H 'accept: application/json' \
>   -H 'Content-Type: application/json' \
>   -d '{
>   "trialId": 1,
>   "numDays": -1
> }'

Check the tasks table in the database to see if the associated tasks for the trial have updated the log retention days to -1.

For WebUI -
1. Create a few experiments.
2. Go to the Experiments list in the Web UI and select a few of them and then click on the actions button.
![EC0E20C0-1338-4317-BDCC-F7F637E8A2D7](https://github.com/determined-ai/determined/assets/131198887/d28f6b56-4d25-4b95-854c-551a7af13866)
3. Select the Retain Logs option from the drop down.
4. Type any number in the Number of days field when prompted.
8. Wait for a success message.
9. Check the tasks table in the database to see if the associated tasks for the experiment have updated the log retention days to the number you put in.
10. Select a few more experiments and follow the above steps, but this time click on the 'Forever' checkbox instead.
11. Wait for a success message.
12.  Check the tasks table in the database to see if the associated tasks for the experiment have updated the log retention days to -1.
13. Also check the Trial List Table to see if the Log Retention Days Column shows up and has the correct days.

## Commentary (optional)





## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.

## Ticket

DET-9995
